File: sdk/javascript/src/instrumentation/mistral.ts
/**
 * MistralInstrumentor — automatic usage capture for the official
 * `@mistralai/mistralai` npm package's `chat.complete`/`chat.stream`.
 *
 *     import { Mistral } from "@mistralai/mistralai";
 *     import { MistralInstrumentor } from "@costorah/sdk";
 *
 *     new MistralInstrumentor().instrument();
 *
 *     const client = new Mistral({ apiKey: "..." });
 *     await client.chat.complete({ model: "mistral-large-latest", messages: [] });
 *
 * `@mistralai/mistralai` doesn't export its `Chat` resource class via any
 * public subpath (Node's package.json `exports` field only whitelists a
 * handful of subpaths, none of them `Chat`) — so unlike OpenAI/Anthropic,
 * this instrumentor can't `require()` the class directly. It reaches the
 * same shared prototype another way: `chat.complete`/`chat.stream` are
 * defined on `Chat.prototype`, which is identical across every `Mistral`
 * client instance. Constructing one throwaway `Mistral` client (no
 * network call — the constructor just wires config) yields a real
 * reference to the same shared prototype every future client uses, so
 * patching it once still intercepts calls made by clients the caller
 * constructs independently afterward.
 */

import { createRequire } from "node:module";

import type { UsageStatus } from "../types.js";
import { BaseInstrumentor, InstrumentationError, makeExtractedUsage } from "./base.js";
import type { ExtractedUsage } from "./base.js";
import { calculateCost } from "./pricing.js";
import { instrumentedAsyncStream } from "./streaming.js";
import { submit } from "./submission.js";

const nodeRequire = createRequire(import.meta.url);

interface MistralUsage {
  prompt_tokens?: number;
  completion_tokens?: number;
  total_tokens?: number;
}

function extract(response: unknown): Record<string, unknown> {
  const usage = (response as { usage?: MistralUsage } | undefined)?.usage;
  if (!usage) return {};
  return {
    inputTokens: usage.prompt_tokens ?? 0,
    outputTokens: usage.completion_tokens ?? 0,
    totalTokens: usage.total_tokens,
  };
}

function aggregateStream(events: unknown[]): Record<string, unknown> {
  for (let i = events.length - 1; i >= 0; i--) {
    const chunk = (events[i] as { data?: unknown }).data;
    const raw = chunk ? extract(chunk) : {};
    if (Object.keys(raw).length > 0) return raw;
  }
  return {};
}

function generateRequestId(): string {
  const random =
    typeof crypto !== "undefined" && "randomUUID" in crypto
      ? crypto.randomUUID().replace(/-/g, "")
      : `${Date.now().toString(16)}${Math.random().toString(16).slice(2)}`;
  return `sdk_js_instr_${random}`;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ChatMethod = (this: any, ...args: unknown[]) => unknown;

export class MistralInstrumentor extends BaseInstrumentor {
  readonly name = "mistral";

  private originalComplete: ChatMethod | undefined;
  private originalStream: ChatMethod | undefined;
  private chatProto: Record<string, unknown> | undefined;

  extractUsage(response: unknown): Record<string, unknown> {
    return extract(response);
  }

  normalize(
    rawUsage: Record<string, unknown>,
    context: { model: string; latencyMs: number; status: UsageStatus; requestId?: string },
  ): ExtractedUsage {
    const inputTokens = Number(rawUsage.inputTokens ?? 0);
    const outputTokens = Number(rawUsage.outputTokens ?? 0);
    const { cost, estimated } = this.calculateCostEnabled
      ? calculateCost("mistral", context.model, inputTokens, outputTokens)
      : { cost: 0, estimated: false };
    const metadata: Record<string, unknown> = {};
    if (this.captureMetadata) metadata.costEstimated = estimated;

    return makeExtractedUsage({
      provider: "mistral",
      model: context.model,
      inputTokens,
      outputTokens,
      totalTokens: rawUsage.totalTokens as number | undefined,
      cost,
      latencyMs: context.latencyMs,
      status: context.status,
      requestId: context.requestId ?? generateRequestId(),
      metadata,
    });
  }

  protected applyPatches(): void {
    let MistralClient: new (opts: { apiKey: string }) => { chat: object };
    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const mod: any = nodeRequire("@mistralai/mistralai");
      MistralClient = mod.Mistral;
      if (!MistralClient) throw new Error("Mistral export not found");
    } catch (err) {
      throw new InstrumentationError(
        "The '@mistralai/mistralai' package is not installed. Install it with " +
          "`npm install @mistralai/mistralai` to use this instrumentor.",
      );
    }

    const probe = new MistralClient({ apiKey: "costorah-instrumentation-probe" });
    const chatProto = Object.getPrototypeOf(probe.chat) as Record<string, unknown>;
    this.chatProto = chatProto;
    this.originalComplete = chatProto.complete as ChatMethod;
    this.originalStream = chatProto.stream as ChatMethod;

    chatProto.complete = this.wrapComplete(this.originalComplete);
    chatProto.stream = this.wrapStream(this.originalStream);
  }

  protected removePatches(): void {
    if (!this.chatProto) return;
    if (this.originalComplete) this.chatProto.complete = this.originalComplete;
    if (this.originalStream) this.chatProto.stream = this.originalStream;
    this.chatProto = undefined;
    this.originalComplete = undefined;
    this.originalStream = undefined;
  }

  private async submitResult(
    model: string,
    raw: Record<string, unknown>,
    latencyMs: number,
    status: UsageStatus,
  ): Promise<void> {
    const usage = this.normalize(raw, { model, latencyMs, status });
    this.recordCaptured();
    await submit(usage);
  }

  private wrapComplete(original: ChatMethod): ChatMethod {
    const submitResult = this.submitResult.bind(this);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return async function (this: any, ...args: unknown[]) {
      const params = (args[0] ?? {}) as { model?: string };
      const model = params.model ?? "unknown";
      const start = Date.now();
      let result: unknown;
      try {
        result = await original.apply(this, args);
      } catch (err) {
        await submitResult(model, {}, Date.now() - start, "error");
        throw err;
      }
      await submitResult(model, extract(result), Date.now() - start, "success");
      return result;
    };
  }

  private wrapStream(original: ChatMethod): ChatMethod {
    const submitResult = this.submitResult.bind(this);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return async function (this: any, ...args: unknown[]) {
      const params = (args[0] ?? {}) as { model?: string };
      const model = params.model ?? "unknown";
      const start = Date.now();
      const result = await original.apply(this, args);
      return instrumentedAsyncStream(
        result as AsyncIterable<unknown>,
        start,
        async (chunks, elapsedMs, error) => {
          const raw = aggregateStream(chunks);
          await submitResult(model, raw, elapsedMs, error ? "error" : "success");
        },
      );
    };
  }
}

File: sdk/python/costorah/instrumentation/mistral.py
"""
MistralInstrumentor — automatic usage capture for the official `mistralai`
Python package's `Chat.complete`/`complete_async`/`stream`/`stream_async`.

    from mistralai import Mistral
    from costorah.instrumentation import MistralInstrumentor

    MistralInstrumentor().instrument()

    client = Mistral(api_key="...")
    client.chat.complete(model="mistral-large-latest", messages=[...])
"""

from __future__ import annotations

import functools
import time
from typing import Any

from costorah._logging import get_logger
from costorah._util import generate_request_id
from costorah.instrumentation._streaming import InstrumentedAsyncStream, InstrumentedSyncStream
from costorah.instrumentation._submission import submit
from costorah.instrumentation.base import BaseInstrumentor, ExtractedUsage, InstrumentationError
from costorah.instrumentation.pricing import calculate_cost
from costorah.types import UsageStatus

_log = get_logger(__name__)


def _extract(response: Any) -> dict[str, Any]:
    usage = getattr(response, "usage", None)
    if usage is None:
        return {}
    return {
        "input_tokens": getattr(usage, "prompt_tokens", 0) or 0,
        "output_tokens": getattr(usage, "completion_tokens", 0) or 0,
        "total_tokens": getattr(usage, "total_tokens", None),
    }


def _aggregate_stream(events: list[Any]) -> dict[str, Any]:
    for event in reversed(events):
        chunk = getattr(event, "data", None)
        if chunk is None:
            continue
        raw = _extract(chunk)
        if raw:
            return raw
    return {}


class MistralInstrumentor(BaseInstrumentor):
    name = "mistral"

    def extract_usage(self, response: Any) -> dict[str, Any]:
        return _extract(response)

    def normalize(
        self,
        raw_usage: dict[str, Any],
        *,
        model: str,
        latency_ms: int,
        status: UsageStatus,
        request_id: str | None = None,
    ) -> ExtractedUsage:
        input_tokens = int(raw_usage.get("input_tokens", 0) or 0)
        output_tokens = int(raw_usage.get("output_tokens", 0) or 0)
        cost, estimated = (
            calculate_cost("mistral", model, input_tokens, output_tokens)
            if self.calculate_cost_enabled
            else (0.0, False)
        )
        metadata: dict[str, Any] = {}
        if self.capture_metadata:
            metadata["cost_estimated"] = estimated
        return ExtractedUsage(
            provider="mistral",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=raw_usage.get("total_tokens"),
            cost=cost,
            latency_ms=latency_ms,
            status=status,
            request_id=request_id or generate_request_id(),
            metadata=metadata,
        )

    def _apply_patches(self) -> None:
        try:
            from mistralai.chat import Chat
        except ImportError as exc:
            raise InstrumentationError(
                "The 'mistralai' package is not installed. Install it with "
                "`pip install mistralai` to use this instrumentor."
            ) from exc

        self._patch(Chat, "complete", self._make_wrapper(Chat.__dict__["complete"], "sync"))
        self._patch(
            Chat, "complete_async", self._make_wrapper(Chat.__dict__["complete_async"], "async")
        )
        self._patch(Chat, "stream", self._make_wrapper(Chat.__dict__["stream"], "sync_stream"))
        self._patch(
            Chat,
            "stream_async",
            self._make_wrapper(Chat.__dict__["stream_async"], "async_stream"),
        )

    def _make_wrapper(self, original: Any, mode: str) -> Any:
        instrumentor = self

        def submit_result(model: str, raw: dict[str, Any], elapsed_ms: int, status: str) -> None:
            usage = instrumentor.normalize(
                raw,
                model=model,
                latency_ms=elapsed_ms,
                status=status,  # type: ignore[arg-type]
            )
            instrumentor._record_captured()
            submit(usage, client=instrumentor._client)

        if mode == "async":

            @functools.wraps(original)
            async def async_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
                model = kwargs.get("model", "unknown")
                start = time.perf_counter()
                try:
                    result = await original(self_client, *args, **kwargs)
                except Exception:
                    submit_result(model, {}, instrumentor._elapsed_ms(start), "error")
                    raise
                submit_result(model, _extract(result), instrumentor._elapsed_ms(start), "success")
                return result

            return async_wrapper

        if mode == "sync_stream":

            @functools.wraps(original)
            def sync_stream_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
                model = kwargs.get("model", "unknown")
                start = time.perf_counter()
                result = original(self_client, *args, **kwargs)

                def on_complete(
                    chunks: list[Any], elapsed_ms: int, error: Exception | None
                ) -> None:
                    raw = _aggregate_stream(chunks)
                    submit_result(model, raw, elapsed_ms, "error" if error else "success")

                return InstrumentedSyncStream(iter(result), start, on_complete)

            return sync_stream_wrapper

        if mode == "async_stream":

            @functools.wraps(original)
            async def async_stream_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
                model = kwargs.get("model", "unknown")
                start = time.perf_counter()
                result = await original(self_client, *args, **kwargs)

                def on_complete(
                    chunks: list[Any], elapsed_ms: int, error: Exception | None
                ) -> None:
                    raw = _aggregate_stream(chunks)
                    submit_result(model, raw, elapsed_ms, "error" if error else "success")

                return InstrumentedAsyncStream(result.__aiter__(), start, on_complete)

            return async_stream_wrapper

        @functools.wraps(original)
        def sync_wrapper(self_client: Any, *args: Any, **kwargs: Any) -> Any:
            model = kwargs.get("model", "unknown")
            start = time.perf_counter()
            try:
                result = original(self_client, *args, **kwargs)
            except Exception:
                submit_result(model, {}, instrumentor._elapsed_ms(start), "error")
                raise
            submit_result(model, _extract(result), instrumentor._elapsed_ms(start), "success")
            return result

        return sync_wrapper
